Added `omit-events` flag to allow disabling of event emission

Vishal-Chdhry opened the pull request:
Signed-off-by: Vishal Choudhary <sendtovishalchoudhary@gmail.com>

## Explanation

Added `omit-events` flag to allow disabling of PolicyApplied events
New PR as the code has been refactored.
<!--
In a couple sentences, explain why this PR is needed and what it addresses. This should be an explanation a non-developer user can understand and covers the "why" question. It should also clearly indicate whether this PR represents an addition, a change, or a fix of existing behavior. This explanation will be used to assist in the release note drafting process.

THIS IS MANDATORY.
-->

## Related issue
Closes: #5215 

<!--
Please link the GitHub issue this pull request resolves in the format of `Closes #1234`. If you discussed this change
with a maintainer, please mention her/him using the `@` syntax (e.g. `@JimBugwadia`).

If this change neither resolves an existing issue nor has sign-off from one of the maintainers, there is a
chance substantial changes will be requested or that the changes will be rejected.

You can discuss changes with maintainers in the [Kyverno Slack Channel](https://kubernetes.slack.com/).
-->

## Milestone of this PR
<!--

Add the milestone label by commenting `/milestone 1.2.3`.

-->

## What type of PR is this

<!--

> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading white spaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
-->

## Proposed Changes

<!--
Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 

***NOTE***: If this PR results in new or altered behavior which is user facing, you **MUST** read and follow the steps outlined in the [PR documentation guide](pr_documentation.md) and add Proof Manifests as defined below.
-->

### Proof Manifests

<!--
Read and follow the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) for more details first. This section is for pasting your YAML manifests (Kubernetes resources and Kyverno policies) and Kyverno CLI test manifests which allow maintainers to prove the intended functionality is achieved by your PR. Please use proper fenced code block formatting, for example:

# Kubernetes resource

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: roles-dictionary
  namespace: default
data:
  allowed-roles: "[\"cluster-admin\", \"cluster-operator\", \"tenant-admin\"]"
```

# Kyverno CLI test manifest (please see docs for latest manifest format at https://kyverno.io/docs/kyverno-cli/). See kyverno/policies for complete examples of all related test files.

```yaml
name: prepend-image-registry
policies:
  - prepend_image_registry.yaml
resources:
  - resource.yaml
variables: values.yaml
results:
  - policy: prepend-registry
    rule: prepend-registry-containers
    resource: mypod
    # if mutate rule
    patchedResource: patchedResource01.yaml
    kind: Pod
    result: pass
```
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
-->

- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [x] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is: https://github.com/kyverno/website/issues/764
  <!-- Uncomment to link to the issue -->
  <!-- https://github.com/kyverno/website/issues/1 -->

## Further Comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution
you did and what alternatives you considered, etc...
-->

Vishal-Chdhry commented on pkg/event/controller.go:
```diff
@@ -96,6 +100,12 @@ func (gen *generator) Add(infos ...Info) {
 			continue
 		}
+		for _, event := range gen.eventsApplied {
+			if info.Reason == Reason(event) {
```
This is the only place where we are testing whether the values passed to the flag match the allowed values `PolicyViolation`, `PolicyApplied`, `PolicyError`, `PolicySkipped`, Is this OK?

Vishal-Chdhry replied:
/cc @realshuting

realshuting commented on cmd/background-controller/main.go:
```diff
@@ -84,10 +85,12 @@ func main() {
 	flagset := flag.NewFlagSet("updaterequest-controller", flag.ExitOnError)
 	flagset.IntVar(&genWorkers, "genWorkers", 10, "Workers for the background controller.")
 	flagset.IntVar(&maxQueuedEvents, "maxQueuedEvents", 1000, "Maximum events to be queued.")
+	flagset.StringVar(&eventsApplied, "eventsApplied", "", "Set this flag to a comma sperated list of PolicyViolation, PolicyApplied, PolicyError, PolicySkipped to disable events, e.g. --emit-events=PolicyApplied,PolicyViolation")
```
Wrong flag name `eventsApplied` here.

realshuting commented on pkg/event/controller.go:
```diff
@@ -96,6 +100,12 @@ func (gen *generator) Add(infos ...Info) {
 		}
+		for _, event := range gen.eventsApplied {
+			if info.Reason == Reason(event) {
+				logger.V(3).Info("skipping policy event applied event creation", "kind", info.Kind, "name", info.Name, "namespace", info.Namespace, "reason", info.Reason)
```
May want to increase the log verbosity to 6. Otherwise we will have flooded logs.

realshuting reviewed (changes requested):
Do we need to add it to the cleanup controller?

Can you add kuttl tests covering difference scenarios?

chipzoller commented:
We should really not be adding flags in a patch release. What is the problem in getting this in 1.10.0?

realshuting commented:
> We should really not be adding flags in a patch release. What is the problem in getting this in 1.10.0?

Adjust the milestone to 1.10.0. @Vishal-Chdhry - we are almost there, can you address the comments?

Vishal-Chdhry commented:
> Can you add kuttl tests covering difference scenarios?

How do I add kuttl change to a cli feature

Vishal-Chdhry commented:
@realshuting

realshuting commented:
> > Can you add kuttl tests covering difference scenarios?
> 
> How do I add kuttl change to a cli feature

The flag is added to the Kyverno controllers, you want configure Kyverno with this flag while testing (or during the initial setup) and use kuttl to check the desired results. Does this make sense?

Vishal-Chdhry commented:
@realshuting we dont have maxQueued events in the cleanup-controller https://github.com/kyverno/kyverno/pull/6072#issuecomment-1408116648

Vishal-Chdhry commented:
@realshuting Something like this on the admission controller?

> The flag is added to the Kyverno controllers, you want configure Kyverno with this flag while testing (or during the initial setup) and use kuttl to check the desired results. Does this make sense?

```yaml
spec:
      containers:
      - args:
        - --emit-events=PolicyApplied,PolicyViolation"
```

realshuting commented:
> 

Yes.

chipzoller commented:
This flag should be subject to the new `features` abstraction provided in the Helm chart, especially since not all controllers receive the flag. (cc @eddycharly )

realshuting commented on test/conformance/kuttl/events/emit-events/01-admission-controller-apply.yaml:
Looks like the testdir has been overridden.
```
2023/05/03 15:22:26 kutt-test config testdirs is overridden with args: [ ./test/conformance/kuttl/events ]
=== RUN   kuttl
    harness.go:494: starting setup
    harness.go:257: running tests using configured kubeconfig.
    harness.go:294: Successful connection to cluster at: https://127.0.0.1:37751
    harness.go:375: running tests
    harness.go:390: step "admission-controller-apply" apply path admission-controlller.yaml: failed to find YAML files in test/conformance/kuttl/events/emit-events/admission-controlller.yaml: file mode issue with stat test/conformance/kuttl/events/emit-events/admission-controlller.yaml: no such file or directory
    harness.go:545: cleaning up
    harness.go:[6](https://github.com/kyverno/kyverno/actions/runs/4873549106/jobs/8693410557?pr=7010#step:9:7)02: removing temp folder: ""
--- FAIL: kuttl (0.32s)
```

@Vishal-Chdhry - can you run the test locally and see if it passes?

Vishal-Chdhry replied:
Hi @realshuting ! I have fixed that typo in the policy.
In the kuttl test, I am trying to 
1. Update the admission controller deployment to include args (which is working when i apply that policy using kubectl)
2. Create a policyApplied event (this is working as well)
3. Create a negative test to check that policyApplied event was not emitted

I am unclear about how to do this, I have created an Event kind with reason set to PolicyApplied, but I can't figure out how to specify that this event wont be emitted, right now I am doing this in the kuttl test step

```yaml
apiVersion: kuttl.dev/v1beta1
kind: TestStep
apply:
- file: event-assert.yaml
  shouldFail: true
error:
- event-assert.yaml
```

realshuting replied:
>Create a policyApplied event (this is working as well)

I don't think you want to create the `policyApplied` type of event directly. instead you want to create a policy and the resource that generates a `policyApplied` event and assert this type of event never gets created.

Does it make sense?

Vishal-Chdhry replied:
Yes that makes sense, I have created a [policy](https://github.com/Vishal-Chdhry/kyverno/blob/policy-applied-flags/test/conformance/kuttl/events/emit-events/policy.yaml) and a [resource](https://github.com/Vishal-Chdhry/kyverno/blob/policy-applied-flags/test/conformance/kuttl/events/emit-events/resource.yaml) which generates the `policyApplied` event. 

(I just realised, that since the flag is named `emit-events` it _should_ emit those events whose types match, I have updated the behavior in the last commit, please review that)

I am updating the test to emit the policyApplied event (instead of _not_ emitting it)

realshuting replied:
This looks good! How about add another negative test to apply a resource that violates the policy and check the policyViolation type of event is not created?

Vishal-Chdhry commented on pkg/event/controller.go:
```diff
@@ -96,7 +100,21 @@ func (gen *generator) Add(infos ...Info) {
 			continue
 		}
-		gen.queue.Add(info)
+		// if the `emit-event` flag is not set
+		// directly add the event to the queue
+		if len(gen.emitEvents) == 0 {
+			gen.queue.Add(info)
+		}
+		// flag is set
+		// only add the event when the reason of event is specified in "emit-events" flag
+		for _, eventReason := range gen.emitEvents {
+			if info.Reason == Reason(eventReason) {
+				logger.V(6).Info("creating event", "kind", info.Kind, "name", info.Name, "namespace", info.Namespace, "reason", info.Reason, "event reason", eventReason)
+				gen.queue.Add(info)
+			} else {
+				logger.V(6).Info("event reason not matched", "kind", info.Kind, "name", info.Name, "namespace", info.Namespace, "reason", info.Reason, "event reason", eventReason)
+			}
+		}
```
This looks like the right logic, if the flag is not set, directly add the event to the queue, otherwise only add the event when the `reason` is also present in the `gen.emitEvents`

chipzoller reviewed (changes requested):
Thanks for adding a kuttl test. However, you're missing a README file which describes the test. Please see the scaffolding directory for an example. Also, because this test requires modifying the Kyverno deployment, we can't put this with the rest of the tests. We'll need to start a separate directory for these and also not include them with the standard tests.

Why don't you create /test/conformance/kuttl/flags and move this test case there (along with adding a README)? We'll figure out how to pick these up later.

Vishal-Chdhry commented:
@chipzoller I will update it! Also, I have to create a "negative" kuttl test, and I had this doubt,

Say I created a policy, and a resource, that will create a `PolicyViolation` event, That event will not be emitted, how to i put that in a teststep that the event wont be emitted, and that is the correct behavior
```yaml
apiVersion: kuttl.dev/v1beta1
kind: TestStep
apply:
- file: event-assert.yaml
  shouldFail: true
error:
- event-assert.yaml
```

chipzoller commented:
I do not know if kuttl can do things like wildcards in assertions/errors for resources. but one thing you could do is use a command definition like shown [here](https://github.com/kyverno/kyverno/blob/main/test/conformance/kuttl/_aaa_template_resources/scripts/02-script-check-for-output.yaml) which just calls `kubectl`. You can write just a simple script, like what's shown to get Events in wherever it should be created and do a simple `grep` or something to ensure it fails and, if so, exit 0.

chipzoller commented:
Since this flag gets applied to 3 out of the 4 controllers, should this be gated behind the Helm chart `feature` object? Seems like it maybe should.

Vishal-Chdhry commented:
@realshuting I am not sure if the negative test is correct or not, can you please review that!

realshuting commented on test/conformance/kuttl/flags/standard/emit-events/05-script.yaml:
```diff
@@ -0,0 +1,10 @@
+kind: TestStep
+commands:
+- script: |
+    if kubectl apply -f resource-fail.yaml 2>&1 | grep -q 'Error from server' 
```
You should apply the resource that results in a violation (in audit mode) and grep the event with `PolicyViolation` type, and exit 1 if it's found, otherwise the test passes.

realshuting commented on test/conformance/kuttl/flags/standard/emit-events/admission-controller.yaml:
```diff
@@ -0,0 +1,170 @@
+          image: "ghcr.io/kyverno/kyverno:latest"
+          imagePullPolicy: IfNotPresent
+          args:
+            - --emit-events=PolicyViolation
```
@Vishal-Chdhry - in your test you are enabling `PolicyViolation` type of events specifically, but why asserting the `PolicyApplied` even in step 4  test/conformance/kuttl/flags/standard/emit-events/04-event.yaml?

I pulled your change locally and applied both the policy and resources, but I couldn't get any Kyverno events:
```
k get event -A  | grep -e policy
```
Am I missing anything?

Vishal-Chdhry replied:
I tested it locally with `PolicyApplied` 😬 in the admission controller yaml, forgot it was `PolicyViolation` here. (but if the kuttl tests passed I might be doing something wrong here)

In my testing when the flag was set to `PolicyApplied` and a `PolicyViolation` event was generated, there was no `PolicyViolation` event in the logs, which means `PolicyViolation` event was not generated

realshuting commented on cmd/background-controller/main.go:
```diff
@@ -84,10 +85,12 @@ func main() {
 	flagset := flag.NewFlagSet("updaterequest-controller", flag.ExitOnError)
 	flagset.IntVar(&genWorkers, "genWorkers", 10, "Workers for the background controller.")
 	flagset.IntVar(&maxQueuedEvents, "maxQueuedEvents", 1000, "Maximum events to be queued.")
+	flagset.StringVar(&emitEvents, "emit-events", "", "Set this flag to a comma sperated list of PolicyViolation, PolicyApplied, PolicyError, PolicySkipped to disable events, e.g. --emit-events=PolicyApplied,PolicyViolation")
```
Is this statement correct?

`emit-events` indicates one wants to generate that type of event, but the description says "to disable events".

Vishal-Chdhry replied:
i think I should change it to omit events

realshuting reviewed (commented):
Can you update the PR description?

>Added eventsApplied flag to allow disabling of PolicyApplied events

This seems to be invalid.

Vishal-Chdhry commented:
@realshuting I have updated the flags to omit events,
this is the output after i have added `--omit-events=PolicyViolation` to the deployment
```bash
» kubectl create -f resource.yaml
configmap/foo created

» kubectl create -f resource-fail.yaml                                                                              
Error from server: error when creating "resource-fail.yaml": admission webhook "validate.kyverno.svc-fail" denied the request: 

policy ConfigMap/default/bar for resource violation: 

require-labels:
  require-team: 'validation error: The label `team` is required. rule require-team
    failed at path /metadata/labels/team/'

» kubectl logs deployment/kyverno-admission-controller -n kyverno | grep "reason=\"PolicyApplied\""
Defaulted container "kyverno" out of: kyverno, kyverno-pre (init)
I0509 16:47:21.909257       1 event.go:307] "Event occurred" object="default/require-labels" fieldPath="" kind="Policy" apiVersion="kyverno.io/v1" type="Normal" reason="PolicyApplied" message="ConfigMap default/foo: pass"

» kubectl logs deployment/kyverno-admission-controller -n kyverno | grep "reason=\"PolicyViolation\""  
Defaulted container "kyverno" out of: kyverno, kyverno-pre (init)   
```

chipzoller commented:
So this is a second container flag?

Vishal-Chdhry commented:
@chipzoller I replaced the `emit-events` flag with `omit-events` which will now _NOT_ emit events if the event type is specified in the flag

chipzoller commented:
Ok, will you please update your corresponding documentation issue since this flag is now different?

realshuting reviewed (approved):
Good job @Vishal-Chdhry !

realshuting reviewed (approved):
Good job @Vishal-Chdhry !

realshuting commented:
/cherry-pick release-1.10

chipzoller commented:
To which controllers is this flag applicable, @Vishal-Chdhry?
